src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+/*******************************************************************\
+
+Module: Generates string constraints for the Java format function
+
+Author: Romain Brenguier
+
+Date:   May 2017
+
+\*******************************************************************/
+
+/// \file
+/// Generates string constraints for the Java format function
+
+#include<iomanip>
+#include<string>
+#include<regex>
+#include<vector>
+
+#include<util/std_expr.h>
+
+#include"string_constraint_generator.h"
+
+// Format specifier describes how a value should be printed.
+class string_constraint_generatort::format_specifiert
+{
+public:
+  // Constants describing the meaning of characters in format specifiers.
+  static const char DECIMAL_INTEGER          ='d';
+  static const char OCTAL_INTEGER            ='o';
+  static const char HEXADECIMAL_INTEGER      ='x';
+  static const char HEXADECIMAL_INTEGER_UPPER='X';
+  static const char SCIENTIFIC               ='e';
+  static const char SCIENTIFIC_UPPER         ='E';
+  static const char GENERAL                  ='g';
+  static const char GENERAL_UPPER            ='G';
+  static const char DECIMAL_FLOAT            ='f';
+  static const char HEXADECIMAL_FLOAT        ='a';
+  static const char HEXADECIMAL_FLOAT_UPPER  ='A';
+  static const char CHARACTER                ='c';
+  static const char CHARACTER_UPPER     ='C';
+  static const char DATE_TIME           ='t';
+  static const char DATE_TIME_UPPER     ='T';
+  static const char BOOLEAN             ='b';
+  static const char BOOLEAN_UPPER       ='B';
+  static const char STRING              ='s';
+  static const char STRING_UPPER        ='S';
+  static const char HASHCODE            ='h';
+  static const char HASHCODE_UPPER      ='H';
+  static const char LINE_SEPARATOR      ='n';
+  static const char PERCENT_SIGN        ='%';
+
+  format_specifiert() { }
+
+  format_specifiert(
+    int _index,
+    std::string _flag,
+    int _width,
+    int _precision,
+    bool _dt,
+    char c):
+      index(_index),
+      flag(_flag),
+      width(_width),
+      precision(_precision),
+      conversion(c)
+  { }
+
+  int index=-1;
```

Comment:
Maybe move declarations before constructor. Also note that Java is incredibly stupid, and the argument index is **1-indexed and not 0-indexed** (https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html).